internal/states/encryption/encryptionconfig/config.go

```diff
@@ -0,0 +1,213 @@
+// Package encryptionconfig contains the data structures and constants to configure client-side state encryption.
+package encryptionconfig
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/opentofu/opentofu/internal/logging"
+)
+
+// Config is a configuration for transparent client-side state encryption.
+//
+// There can be more than one configuration with different encryption or key derivation methods.
+type Config struct {
+	KeyProvider KeyProviderConfig `json:"key_provider"`
+
+	Method EncryptionMethodConfig `json:"method"`
+
+	// Required forces encryption operations to fail if they would
+	// result in an unencrypted output.
+	Required bool `json:"required"`
+}
+
+func (c Config) Validate() error {
+	if err := c.KeyProvider.Validate(); err != nil {
+		return err
+	}
+	if err := c.Method.Validate(); err != nil {
+		return err
+	}
+	return nil
+}
+
+const (
+	ConfigKeyDefault   = "default"   // a default for all remote states (backend + any remote_state data sources)
+	ConfigKeyBackend   = "backend"   // for our own remote state backend
+	ConfigKeyStatefile = "statefile" // for local state files
+	ConfigKeyPlanfile  = "planfile"  // for plan files
+)
+
+// ConfigEnvJsonStructure is used to parse multiple named configurations from environment variables.
+//
+// The key either specifies a resource that accesses remote state, such as "terraform_remote_state.foo", or
+// it is one of the pre-defined convenience keys ConfigKeyDefault, ConfigKeyBackend, ConfigKeyStatefile, or
+// ConfigKeyPlanfile.
+//
+// If the key is "terraform_remote_state.foo", its value sets/overrides encryption configuration for
+//
+//	data "terraform_remote_state" "foo" {}
+//
+// For enumerated resources, the format is "terraform_remote_state.foo[17]" or "terraform_remote_state.foo[key]"
+// (no quotes around the for_each key).
+type ConfigEnvJsonStructure map[string]Config
+
+// ConfigEnvName is the name of the environment variable used to configure encryption and decryption
+//
+// Set this environment variable to a json representation of ConfigEnvJsonStructure, or leave it unset/blank
```

Comment:
If we switched to environment variables, this would look as follows:

```
TF_STATE_ENCRYPTION_ENFORCED=1
TF_STATE_ENCRYPTION_KEY_PROVIDER_NAME=bla
TF_STATE_ENCRYPTION_KEY_PROVIDER_SOMECONFIG=somevalue
TF_STATE_ENCRYPTION_METHOD_NAME=bla
TF_STATE_ENCRYPTION_METHOD_SOMECONFIG=somevalue
```

This is not even taking into account separate configuration for remote state data sources. Btw. JSON in env vars was decided in the RFC, if we change this we should update the RFC too.